docs/topic/cluster-config.rst

```diff
@@ -0,0 +1,146 @@
+configuration of our kubernetes cluster. The following command represents
+the currently favored configuration.
+
+.. code:: bash
+
+   gcloud container clusters create
+        --enable-ip-alias \
+        --enable-autoscaling \
+        --max-nodes=20 --min-nodes=2 \
+        --region=us-central1 --node-locations=us-central1-b \
+        --image-type=ubuntu \
+        --disk-size=100 --disk-type=pd-standard \
+        --machine-type=n1-highmem-8 \
+        --cluster-version latest \
+        <cluster-name>
+
+We will try explain the various arguments to this
+
+IP Aliasing
+-----------
+
+``--enable-ip-alias`` creates `VPC Native Clusters <https://cloud.google.com/kubernetes-engine/docs/how-to/alias-ips>`_.
+
+This becomes the default soon, and can be removed once it is the default.
+
+Autoscaling
+-----------
+
+We use the `kubernetes cluster autoscaler <https://cloud.google.com/kubernetes-engine/docs/concepts/cluster-autoscaler>`_
+to scale our node count up and down based on demand. It waits until the cluster is completely full
+before triggering creation of a new node - but that's ok, since new node creation time on GKE is
+pretty quick.
+
+``--enable-autoscaling`` turns the cluster autoscaler on. 
+
+``--min-nodes`` sets the minimum number of nodes that will be maintained
+regardless of demand. This should ideally be 2, to give us some headroom for
+quick starts without requiring scale ups when the cluster is completely empty.
+
+``--max-nodes`` sets the maximum number of nodes that the cluster autoscaler
+will use - this sets the maximum number of concurrent users we can support.
+This should be set to a reasonably high number, but not too high - to protect
+against runaway creation of hundreds of VMs that might drain all our credits
+due to accident or security breach.
+
+Highly available master
+-----------------------
+
+The kubernetes cluster's master nodes are managed by Google Cloud automatically.
+By default, it is deployed in a non-highly-available configuration - only one
+node. This means that upgrades and master configuration changes cause a few minutes
+of downtime for the kubernetes API, causing new user server starts / stops to fail.
+
+We request our cluster masters to have `highly available masters <https://cloud.google.com/kubernetes-engine/docs/concepts/regional-clusters>`_
+with ``--region`` parameter. This specifies the region where our 3 master nodes
+will be spread across in different zones. It costs us nothing extra, so we should
+always do it.
+
+By default, asking for highly available masters also asks for 3x the node count,
+spread across multiple zones. We don't want that, since all our user pods have
+in-memory state & can't be relocated. Specifying ``--node-locations`` explicitly
+lets us control how many and which zones the nodes are located in.
+
+Region / Zone selection
+-----------------------
+
+We generally use the ``us-central1`` region and a zone in it for our clusters -
```

Comment:
Is it worth anticipating latency questions here? I don't think it has mattered for us.